test/addons/e10s-remote/main.js

```diff
@@ -0,0 +1,581 @@
+const { getTabs, openTab, closeTab, setTabURL, getBrowserForTab, getURI } = require('sdk/tabs/utils');
+const { getMostRecentBrowserWindow } = require('sdk/window/utils');
+const { cleanUI } = require("sdk/test/utils");
+const { setTimeout } = require("sdk/timers");
+
+const { Cu } = require('chrome');
+const { async } = Cu.import('resource://gre/modules/Task.jsm', {}).Task;
+
+const { set } = require('sdk/preferences/service');
+const mainWindow = getMostRecentBrowserWindow();
+const isE10S = mainWindow.gMultiProcessBrowser;
+
+// The hidden preload browser messes up our frame counts
+set('browser.newtab.preload', false);
+
+function promiseEvent(emitter, event) {
+  console.log("Waiting for " + event);
+  return new Promise(resolve => {
+    emitter.once(event, (...args) => {
+      console.log("Saw " + event);
+      resolve(args);
+    });
+  });
+}
+
+function promiseDOMEvent(target, event, isCapturing = false) {
+  console.log("Waiting for " + event);
+  return new Promise(resolve => {
+    let listener = (event) => {
+      target.removeEventListener(event, listener, isCapturing);
+      resolve(event);
+    };
+    target.addEventListener(event, listener, isCapturing);
+  })
+}
+
+promiseEventOnItemAndContainer = async(function*(assert, itemport, container, event, item = itemport) {
+  let itemEvent = promiseEvent(itemport, event);
+  let containerEvent = promiseEvent(container, event);
+
+  let itemArgs = yield itemEvent;
+  let [target, ...containerArgs] = yield containerEvent;
+
+  assert.equal(target, item, "Should have seen a container event for the right item");
+  assert.equal(JSON.stringify(itemArgs), JSON.stringify(containerArgs), "Arguments should have matched");
+
+  return itemArgs;
+});
+
+let waitForProcesses = async(function*(loader) {
+  console.log("Starting remote");
+  let { processes, frames, remoteRequire } = loader.require('sdk/remote/parent');
+  remoteRequire(REMOTE_MODULE, module);
+
+  let events = [];
+
+  // In e10s we should expect to see two processes
+  let expectedCount = isE10S ? 2 : 1;
+
+  yield new Promise(resolve => {
+    let count = 0;
+
+    // Wait for a process to be detected
+    let listener = process => {
+      console.log("Saw a process attach");
+      // Wait for the remote module to load in this process
+      process.port.once('sdk/test/load', () => {
+        console.log("Saw a remote module load");
+        count++;
+        if (count == expectedCount) {
+          processes.off('attach', listener);
+          resolve();
+        }
+      });
+    }
+    processes.on('attach', listener);
+  });
+
+  console.log("Remote ready");
+  return { processes, frames, remoteRequire };
+});
+
+if (isE10S) {
+  console.log("Testing in E10S mode");
+  // We expect a child process to already be present, make sure that is the case
+  mainWindow.XULBrowserWindow.forceInitialBrowserRemote();
+
+  // Check that we see a process stop and start
+  exports["test process restart"] = function*(assert) {
```

Comment:
just have this function return early if `isE10S` is true, to avoid indenting the whole test.